src/client.rs

```diff
@@ -285,71 +280,181 @@ impl Client {
+    ///
+    /// # Behaviour
+    /// * If stored data is up-to-date and signature of local data valid, then return local content;
+    /// * Otherwise fetch content from server, merge with local content, verify signature, and return records;
+    ///
+    /// # Errors
+    /// If an error occurs while fetching or verifying records, a [`ClientError`] is returned.
+    pub fn sync<T>(&mut self, expected: T) -> Result<Collection, ClientError>
+    where
+        T: Into<Option<u64>>,
+    {
+        let storage_key = format!("{}/{}:collection", self.bucket_name, self.collection_name);
+
+        debug!("Retrieve from storage with key={:?}", storage_key);
+        let stored_bytes: Vec<u8> = self
+            .storage
+            .retrieve(&storage_key)
+            .unwrap_or(None)
+            .unwrap_or_else(Vec::new);
         let stored: Option<Collection> = serde_json::from_slice(&stored_bytes).unwrap_or(None);
 
-        // Note: when we implement the `sync()` method in #28,
-        // we will remove this, and `.get()` will return the current content of cache if it
-        // is not empty.
-        let remote_timestamp = get_latest_change_timestamp(
-            &self.server_url,
-            &self.bucket_name,
-            &self.collection_name,
-        )?;
+        let remote_timestamp = match expected.into() {
+            Some(v) => v,
+            None => {
+                debug!("Obtain current timestamp.");
+                get_latest_change_timestamp(
+                    &self.server_url,
+                    &self.bucket_name,
+                    &self.collection_name,
+                )?
+            }
+        };
 
-        if let Some(collection) = stored {
+        if let Some(ref collection) = stored {
             let up_to_date = collection.timestamp == remote_timestamp;
             if up_to_date && self.verifier.verify(&collection).is_ok() {
                 debug!("Local data is up-to-date and valid.");
-                return Ok(collection.records);
+                return Ok(collection.to_owned());
             }
         }
 
         info!("Local data is empty, outdated, or has been tampered. Fetch from server.");
-        let collection = self._fetch_records_and_verify(remote_timestamp)?;
+        let (local_records, local_timestamp) = match stored {
+            Some(c) => (c.records, Some(c.timestamp)),
+            None => (Vec::new(), None),
+        };
+
+        let changeset = get_changeset(
+            &self.server_url,
+            &self.bucket_name,
+            &self.collection_name,
+            Some(remote_timestamp),
+            local_timestamp,
+        )?;
+
+        debug!(
+            "Apply {} changes to {} local records",
+            changeset.changes.len(),
+            local_records.len()
+        );
+        let merged = merge_changes(local_records.to_vec(), changeset.changes);
+
+        let collection = Collection {
+            bid: self.bucket_name.to_owned(),
```

Comment:
I think it's a lot more common to use `clone()` in cases like this, since it captures the intention better. You already have an owned string, and you want to clone it. (This is a purely stylistic difference. The resulting code is identical.)